### PR TITLE
(fix) KH-141: Remote Select dropdown at the end of the form content should display above all content.

### PR DESCRIPTION
**Pull request by hadijahkyampeire**
## Requirements

- [x] This PR has a title that briefly describes the work done, including the ticket number if there is a ticket.
- [x] My work conforms to the [OpenMRS 3.0 Styleguide](https://om.rs/styleguide) and [design documentation](https://zeroheight.com/23a080e38/p/880723-introduction).
- [ ] My work includes tests or is validated by existing tests.

## Summary
- This PR fixes `ng-select` which is inside a `remote-select` and displays the dropdown panel above all content.

## Screenshots
- Before(Ngx-Formentry)
Screenshot 2023-04-11 at 22 29 33
- Dev3(Patient chart)
Screenshot 2023-04-14 at 02 16 53 (2)

- Fixed(Ngx-Formentry)
Screenshot 2023-04-11 at 22 28 39

- Fixed by this PR(Patient chart)
Screenshot 2023-04-14 at 02 13 18 (2)

## Related Issue
- https://issues.openmrs.org/browse/KH-143

## Other
- After merging this we need to test it in the patient chart forms and see if it actually works.

### CODE REVIEW

**Comment by hadijahkyampeire** [comment]
@enyachoke for some reason I can't get your name as a reviewer 🤔 , but this is the fix I got hopefully it will fix the issue in the patient chart. cc @rbuisson

**Comment by rbuisson** [comment]
@enyachoke  is off today. @ibacher would you be able to take a look at this today?

File: projects/ngx-formentry/src/components/ngx-remote-select/remote-select.component.html [hunk 1]
```diff
@@ -13,6 +13,7 @@
     typeToSearchText="Please enter 2 or more characters"
     [typeahead]="remoteOptionInput$"
     [(ngModel)]="selectedRemoteOptions"
+    [appendTo]="'body'"
```
[the remark this answers is not in the document]
Comment (hadijahkyampeire) [reply projects/ngx-formentry/src/components/ngx-remote-select/remote-select.component.html:16]: Yah, that's what the docs had, I will test with double quotes only and see. My thinking was that we are appending a string body to the select.

[the remark this answers is not in the document]
Comment (ibacher) [reply projects/ngx-formentry/src/components/ngx-remote-select/remote-select.component.html:16]: I don't think `body` is the right selector for this. It works in the self-contained app here, but that's because the app only displays the form. In the form-engine context, however, this is almost certainly wrong. In the context of the form engine on dev3, it probably makes more sense to have this as `appendTo="my-app-fe-wrapper form"`, i.e., to the form element itself.

File: projects/ngx-formentry/src/components/ngx-remote-select/ngx-remote-select.component.ts [hunk 2]
```diff
@@ -65,6 +66,13 @@ export class RemoteSelectComponent implements OnInit, ControlValueAccessor {
 
   ngOnInit() {
     this.loadOptions();
+    // check if there is enough space at the bottom of the ng-select component
+    const selectElement = document.querySelector('ng-select');
+    const selectBoundingRect = selectElement.getBoundingClientRect();
+    const bottomSpace = window.innerHeight - selectBoundingRect.bottom;
```
Comment (ibacher) [inline projects/ngx-formentry/src/components/ngx-remote-select/ngx-remote-select.component.ts:72]: Outside of the context of the test app, I'm not sure `window.innerHeight` is the right property, since it assumes the form is the full height of the window. It seems like it would be better to compare the `bottom` of the bounding rectangle for the select element with the bottom for the bounding rectangle of the `my-app-root` element (since that's the root of the form tree). That way, if the form engine is rendered in a confined space, it will still continue to work.

**Review by ibacher — CHANGES_REQUESTED** [review]
Thanks @hadijahkyampeire! I think the solution here is generally good, but needs a few tweaks to actually work if this were actually run on dev3. It would actually be quite useful to have a screenshot of that (you can use `yarn link` to setup the forms app to be linked to your local copy of openmrs-ngx-formentry).

**Comment by hadijahkyampeire** [comment]
Thanks @ibacher for the review, I will implement the requested changes and link with the RefApp so that I confirm that it actually works.

**Comment by hadijahkyampeire** [comment]
@ibacher the `yarn link @openmrs/ngx-formentry` command is failing to work in patient chart, I am doing the right thing?

**Comment by hadijahkyampeire** [comment]
@ibacher I believe this work even though I am not able to test it with a select in the patient chart but I was able to see the selectors that appear in patient chart but they are slightly different from the ones in the form engine and this makes the form engine fail for example what is `my-app-root` in patient chart is `app-root` in the form engine And `my-app-fe-wrapper` only appears in the patient chart but not in this repo.

Is there a way to make the selects work well for both repos?

**Comment by hadijahkyampeire** [comment]
I have added a condition that will make it work for both the patient chart and this repo. And I have just appendedTo `form` since both side have the form tag.

**Comment by hadijahkyampeire** [comment]
I have added a condition that will make it work for both the patient chart and this repo. And I have just appendedTo `form` since both side have the form tag.

**Comment by hadijahkyampeire** [comment]
I have added a condition that will make it work for both the patient chart and this repo. And I have just appendedTo `form` since both side have the form tag.

**Comment by hadijahkyampeire** [comment]
@ibacher I have pushed the changes and also tested with the patient chart, will you please take another look? Thanks.

**Review by ibacher — COMMENTED** [review]
Ok, so fundamentally my problem with merging this in is I just don't see how all the code here actually functions. Specifically:

1. In the template we reference a bound property called `appendTo`. This is part of the `ng-select` library API and controls where the options renders. This is always set (in this PR) to "form". Since the element is already a part of the form that we're rendering, it's not quite clear to me why we need this.
2. There's a bunch of code added that computes various bounding rectangles and then updates a property called `appendToParentElement`. However, it is not an `@Input()` or `@Output()` element, is not actually _read_ at any point in the code and does not seem to be an implicit property of the Angular component, so I'm not sure what all that code actually does besides... running some calculations and updating a property. That property isn't rendered anywhere AFAICT, nor does it seem to affect the positioning of any elements, etc.
3. There's also a CSS change, but I don't quite get why it's written as two separate rules, i.e., under what circumstances does the component `.ng-select > .ng-dropdown-panel` exist in this change?

Basically, it seems like the functional part of this PR is adding the `appendTo` property and adding the `margin-top: -1rem;` to `.ng-dropdown-panel` and all the rest of this seems unnecessary. So can we just reduce this PR to that?

Also, in this case, the value we're trying to fix via the `margin-top` thing is set in `px` rather than `rem` [see here](https://github.com/ng-select/ng-select/blob/1497f19df6c31580b90cb3780f005aaf2b278eed/src/ng-select/lib/ng-dropdown-panel.component.ts#L417), so it would be best to express the override in `px` as well.

File: projects/ngx-formentry/src/components/ngx-remote-select/remote-select.component.html [hunk 3]
```diff
@@ -13,6 +13,7 @@
     typeToSearchText="Please enter 2 or more characters"
     [typeahead]="remoteOptionInput$"
     [(ngModel)]="selectedRemoteOptions"
+    [appendTo]="'form'"
```
[the remark this answers is not in the document]
Comment (hadijahkyampeire) [reply projects/ngx-formentry/src/components/ngx-remote-select/remote-select.component.html:16]: Without this append, the dropdown does not show on top of the other content, I tried removing it and the dropdown was appearing behind the form. That's the solution I got that will solve the issue.

[the remark this answers is not in the document]
Comment (hadijahkyampeire) [reply projects/ngx-formentry/src/components/ngx-remote-select/remote-select.component.html:16]: To answer you elaborately and why I used the `appendTo`, 
```
The [appendTo] attribute on an ng-select directive is used to 
specify where the dropdown menu of the select should be 
appended to in the DOM. By default, the dropdown menu is appended to the body of the document, 
which means it will appear on top of all other content on the page.

However, in some cases, you may want to append the dropdown menu 
to a specific element in the DOM, such as a form or a container div. 
This is where the [appendTo] attribute comes in.

When you set [appendTo] to 'form', you are telling the ng-select 
directive to append the dropdown menu to the nearest ancestor 
form element in the DOM. 
This can be useful if you want to ensure that the dropdown menu i
s always contained within the form element, and doesn't appear outside of it.
```

And this is the solution I got to ensure that our dropdown panel appears above the content.
But I am open to a better solution. cc @ibacher

[the remark this answers is not in the document]
Comment (ibacher) [reply projects/ngx-formentry/src/components/ngx-remote-select/remote-select.component.html:16]: Yes, but see the larger note I wrote as a summary. _If_ this is the part doing the work (and it is), then we don't need the code changes in `ngx-remote-select.component.ts`. However, that also means that this isn't _just_ fixing drop-downs at the end of the form. Now _every_ remote-select component will render the drop-down elements as a sibling of the `form` element. That also means that in the css file, we only really need _one_ rule (because there is never a component that will match `.ng-select > .ng-dropdown-panel`). So basically, this PR does too many things, most of which aren't actually used.

An alternative implementation would be to drive this `appendTo` property from some kind of property in the component, in which case the code in `ngx-remote-select.component.ts` and the css file would make sense, but then this property would need to point to the component property and _not_ a fixed selector.

File: README.md [hunk 4]
```diff
@@ -92,6 +92,28 @@ Which is should be fine if your are not working on custom components (Make sure
...
+
+On the patient chart repo cd into `esm-form-entry-app`
+
+While there, run `yarn link` + the path you copied earlier e.g `/Users/hadijahkyampeire/Desktop/openmrs-ngx-formentry/dist/ngx-formentry`.
+
+The above command should install the local copy of the form entry library into the `esm-form-entry-app` package. You can check that this worked by opening the `package.json` manifest file entry for `@openmrs/ngx-formentry`.
+
+When you start the form entry app, it should now be using the local copy of `openmrs-ngx-formentry`
+
+- The key thing to note is that an override for the form entry import map URL is already set up on the dev server.
```
Comment (ibacher) [inline README.md:111]:
```suggestion
The key thing to note is that an override for the form entry import map URL is already set up on the dev server.
```

File: projects/ngx-formentry/src/components/ngx-remote-select/remote-select.component.html (hunk 3 above)
[the remark this answers is not in the document]
Comment (hadijahkyampeire) [reply projects/ngx-formentry/src/components/ngx-remote-select/remote-select.component.html:16]: Actually, the other selects which are not remote get changed by the negative margin I add to the remote select and they look weird, that's why I also need to target their dropdown panel with a 0 margin-top.

[the remark this answers is not in the document]
Comment (hadijahkyampeire) [reply projects/ngx-formentry/src/components/ngx-remote-select/remote-select.component.html:16]: @ibacher I have removed all the implementation in the `ngx-remote-select.component.ts`, the dropdowns are fixed in the `ngx-form-entry` forms but they are not fixed in the patient chart. I am not sure why the fix is not working for the patient chart 🤔 .

[the remark this answers is not in the document]
Comment (hadijahkyampeire) [reply projects/ngx-formentry/src/components/ngx-remote-select/remote-select.component.html:16]: I just realised the the `appendTo` is not reflecting in the patient chart but if I add it manually the dropdown works, maybe the linking is the problem.

[the remark this answers is not in the document]
Comment (hadijahkyampeire) [reply projects/ngx-formentry/src/components/ngx-remote-select/remote-select.component.html:16]: @ibacher @rbuisson I have tested the just the `appendTo` change and it is also working for the patient chart.

**Comment by ibacher** [comment]
Thanks @hadijahkyampeire!

**Comment by hadijahkyampeire** [comment]
Thank you @ibacher for the guidance, now I need to follow this up with a PR to bump the `ngx-formentry` version in the patient chart, right? So that these changes take effect.
